### DISCUSSION TITLE
OpenMP parallelization of energy

### DISCUSSION THREAD

**Discussion by directysj** [mlund/faunus#364] [Show and tell]
Hello @mlund 

I would like to simulate really large system (approximately 20000 particles). Therefore, I made parallel energy.cpp of Faunus, but when running the code. It is using a single thread. It would be really helpful for your suggestions in the code, providing below section wise. 

void PolicyIonIon::updateBox(EwaldData &d, const Point &box) const {
    assert(d.policy == EwaldData::PBC or d.policy == EwaldData::PBCEigen);
    d.box_length = box;
    int n_cutoff_ceil = ceil(d.n_cutoff);
    d.check_k2_zero = 0.1 * std::pow(2 * pc::pi / d.box_length.maxCoeff(), 2);
    int k_vector_size = (2 * n_cutoff_ceil + 1) * (2 * n_cutoff_ceil + 1) * (2 * n_cutoff_ceil + 1) - 1;
    if (k_vector_size == 0) {
        d.k_vectors.resize(3, 1);
        d.Aks.resize(1);
        d.k_vectors.col(0) = Point(1, 0, 0); // Just so it is not the zero-vector
        d.Aks[0] = 0;
        d.num_kvectors = 1;
        d.Q_ion.resize(1);
        d.Q_dipole.resize(1);
    } else {
        double nc2 = d.n_cutoff * d.n_cutoff;
        d.k_vectors.resize(3, k_vector_size);
        d.Aks.resize(k_vector_size);
        d.num_kvectors = 0;
        d.k_vectors.setZero();
        d.Aks.setZero();
        int start_value = 1;

				double factor,dnx2,dny2,k2;
				int ny,nz;

        #pragma omp parallel for firstprivate(nc2,start_value,n_cutoff_ceil)\
				private(ny,nz,factor,dnx2,dny2,k2) schedule(dynamic)
        
				for (int nx = 0; nx <= n_cutoff_ceil; nx++) {
            dnx2 = double(nx * nx);
            factor = (nx > 0) ? 2.0 : 1.0; // optimization of PBC Ewald (and
                                                  // always the case for IPBC Ewald)
            //for (int ny = -n_cutoff_ceil * start_value; ny <= n_cutoff_ceil; ny++) {
            for ( ny = -n_cutoff_ceil * start_value; ny <= n_cutoff_ceil; ny++) {
                dny2 = double(ny * ny);
                for ( nz = -n_cutoff_ceil * start_value; nz <= n_cutoff_ceil; nz++) {
                    Point kv = 2 * pc::pi * Point(nx, ny, nz).cwiseQuotient(d.box_length);
                    k2 = kv.squaredNorm() + d.kappa_squared; // last term is only for Yukawa-Ewald
                    if (k2 < d.check_k2_zero)                       // Check if k2 != 0
                        continue;
                    if (d.use_spherical_sum) {
                        double dnz2 = double(nz * nz);
                        if ((dnx2 + dny2 + dnz2) / nc2 > 1)
                            continue;
                    }
                    d.k_vectors.col(d.num_kvectors) = kv;
                    d.Aks[d.num_kvectors] = factor * exp(-k2 / (4 * d.alpha * d.alpha)) / k2;
                    d.num_kvectors++;
                }
            }
        }
        d.Q_ion.resize(d.num_kvectors);
        d.Q_dipole.resize(d.num_kvectors);
        d.Aks.conservativeResize(d.num_kvectors);
        d.k_vectors.conservativeResize(3, d.num_kvectors);
    }
}

void PolicyIonIon::updateComplex(EwaldData &data, Space::Tgvec &groups) const {

	double qr;

#pragma omp parallel for private(qr) schedule(dynamic)

    for (int k = 0; k < data.k_vectors.cols(); k++) {
        const Point &q = data.k_vectors.col(k);
        EwaldData::Tcomplex Q(0, 0);
        for (auto &g : groups) {       // loop over molecules
            for (auto &particle : g) { // loop over active particles
                //double qr = q.dot(particle.pos);
                qr = q.dot(particle.pos);
                Q += particle.charge * EwaldData::Tcomplex(std::cos(qr),
                                                           std::sin(qr)); // 'Q^q', see eq. 25 in ref.
            }
            data.Q_ion[k] = Q;
        }
    }
}

void PolicyIonIon::updateComplex(EwaldData &d, Change &change, Space::Tgvec &groups, Space::Tgvec &oldgroups) const {
    assert(groups.size() == oldgroups.size());

#pragma omp parallel for schedule(dynamic) 

    for (int k = 0; k < d.k_vectors.cols(); k++) {
        auto &Q = d.Q_ion[k];
        const Point &q = d.k_vectors.col(k);

        for (auto &changed_group : change.groups) {
            auto &g_new = groups.at(changed_group.index);
            auto &g_old = oldgroups.at(changed_group.index);
            for (auto i : changed_group.atoms) {
                if (i < g_new.size()) {
                    double qr = q.dot(g_new[i].pos);
                    Q += g_new[i].charge * EwaldData::Tcomplex(std::cos(qr), std::sin(qr));
                }
                if (i < g_old.size()) {
                    double qr = q.dot(g_old[i].pos);
                    Q -= g_old[i].charge * EwaldData::Tcomplex(std::cos(qr), std::sin(qr));
                }
            }
        }
    }
}

void PolicyIonIonIPBC::updateBox(EwaldData &data, const Point &box) const {
    assert(data.policy == EwaldData::IPBC or data.policy == EwaldData::IPBCEigen);
    data.box_length = box;
    int ncc = std::ceil(data.n_cutoff);
    data.check_k2_zero = 0.1 * std::pow(2 * pc::pi / data.box_length.maxCoeff(), 2);
    int k_vector_size = (2 * ncc + 1) * (2 * ncc + 1) * (2 * ncc + 1) - 1;
    if (k_vector_size == 0) {
        data.k_vectors.resize(3, 1);
        data.Aks.resize(1);
        data.k_vectors.col(0) = Point(1, 0, 0); // Just so it is not the zero-vector
        data.Aks[0] = 0;
        data.num_kvectors = 1;
        data.Q_ion.resize(1);
        data.Q_dipole.resize(1);
    } else {
        double nc2 = data.n_cutoff * data.n_cutoff;
        data.k_vectors.resize(3, k_vector_size);
        data.Aks.resize(k_vector_size);
        data.num_kvectors = 0;
        data.k_vectors.setZero();
        data.Aks.setZero();
        int start_value = 0;

				double xfactor,yfactor,factor;
				double dnx2,dny2,dnz2,k2;

        #pragma omp parallel for firstprivate(nc2,start_value)\
				private(xfactor,yfactor,factor,dnx2,dny2,dnz2,k2) schedule(dynamic)

        for (int nx = 0; nx <= ncc; nx++) {
            dnx2 = double(nx * nx);
            xfactor = (nx > 0) ? 2.0 : 1.0; // optimization of PBC Ewald
            for (int ny = -ncc * start_value; ny <= ncc; ny++) {
                dny2 = double(ny * ny);
                yfactor = (ny > 0) ? 2.0 : 1.0; // optimization of PBC Ewald
                for (int nz = -ncc * start_value; nz <= ncc; nz++) {
                    factor = xfactor * yfactor;
                    if (nz > 0)
                        factor *= 2;
                    Point kv = 2 * pc::pi * Point(nx, ny, nz).cwiseQuotient(data.box_length);
                    k2 = kv.squaredNorm() + data.kappa_squared; // last term is only for Yukawa-Ewald
                    if (k2 < data.check_k2_zero)                       // Check if k2 != 0
                        continue;
                    if (data.use_spherical_sum) {
                        dnz2 = double(nz * nz);
                        if ((dnx2 + dny2 + dnz2) / nc2 > 1)
                            continue;
                    }
                    data.k_vectors.col(data.num_kvectors) = kv;
                    data.Aks[data.num_kvectors] = factor * exp(-k2 / (4 * data.alpha * data.alpha)) / k2;
                    data.num_kvectors++;
                }
            }
        }
        data.Q_ion.resize(data.num_kvectors);
        data.Q_dipole.resize(data.num_kvectors);
        data.Aks.conservativeResize(data.num_kvectors);
        data.k_vectors.conservativeResize(3, data.num_kvectors);
    }
}

void PolicyIonIonIPBC::updateComplex(EwaldData &d, Space::Tgvec &groups) const {
    assert(d.policy == EwaldData::IPBC or d.policy == EwaldData::IPBCEigen);

    #pragma omp parallel for schedule(dynamic)

    for (int k = 0; k < d.k_vectors.cols(); k++) {
        const Point &q = d.k_vectors.col(k);
        EwaldData::Tcomplex Q(0, 0);
        for (auto &g : groups) {
            for (auto &particle : g) {
                Q += q.cwiseProduct(particle.pos).array().cos().prod() * particle.charge; // see eq. 2 in doi:10/css8
            }
        }
        d.Q_ion[k] = Q;
    }
}

void PolicyIonIonIPBC::updateComplex(EwaldData &d, Change &change, Space::Tgvec &groups,
                                     Space::Tgvec &oldgroups) const {
    assert(d.policy == EwaldData::IPBC or d.policy == EwaldData::IPBCEigen);
    assert(groups.size() == oldgroups.size());

#pragma omp parallel for schedule(dynamic)

    for (int k = 0; k < d.k_vectors.cols(); k++) {
        auto &Q = d.Q_ion[k];
        const Point &q = d.k_vectors.col(k);
        for (auto &changed_group : change.groups) {
            auto &g_new = groups.at(changed_group.index);
            auto &g_old = oldgroups.at(changed_group.index);
            for (auto i : changed_group.atoms) {
                if (i < g_new.size())
                    Q += q.cwiseProduct(g_new[i].pos).array().cos().prod() * g_new[i].charge;
                if (i < g_old.size())
                    Q -= q.cwiseProduct(g_old[i].pos).array().cos().prod() * g_old[i].charge;
            }
        }
    }
}
`

**Comment by mlund** [comment]
Thanks for the suggested OpenMP additions to the Ewald code. I wonder if you could provide a bit more detail about the system you want to simulate? My immediate thought is that the bottle-neck is the real-space part, i.e. `Energy::Nonbonded`. Could you try to profile this, for example by running your system with and without Ewald summation? (coulomb type `plain` vs. `ewald`) Another major speed increase would likely come from a cell-list which is currently in the works -- see pull-request (#346).

**Reply by directysj** [reply]
Thanks a lot for your suggestions. In my system, there are several proteins where interactions are bonding as well as non-bonding -- Lennard-Jones and Coulomb type .  I will compare the system as per your suggestions.

**Reply by mlund** [reply]
OK, how is the solvent handled? If it is implicit and high dielectric, I would probably not use ewald summation. Another tip is to use use [mass center cutoffs](https://faunus.readthedocs.io/en/latest/_docs/energy.html#mass-center-cutoffs).

**Reply by directysj** [reply]
Yes, (water) solvent is treated implicitly.

**Reply by mlund** [reply]
> Yes, (water) solvent is treated implicitly.

OK, and how about salt -- is that also implicit? For systems this large, I think having explicit salt may be very difficult to sample as macromolecular moves can only be very small in order not to bump into salt particles / counter ions. If you don't have strong ion-specific effects or multivalent ions, I suggest using `yukawa` for electrostatics. If your particles tend to cluster, you could also enhance sampling using the `cluster` move.

**Reply by directysj** [reply]
Thanks a lot. I agree with you for explicit ions. Salt is also treated implicitly, only counter ions to make system electrically neutral. I will try with cluster move. Still, I think it is needed to make parallel the non-bonded interactions, i.e., Lennard-Jones and Coulomb. I will let you know with its testing.

**Reply by mlund** [reply]
In the implicit salt framework (Poisson-Boltzmann approximation; `yukawa`) it's normal to have a system net-charge as it's neutralized by implicit counterions, so there's no need to add counter-ions, unless you think their surface correlations are important. We have simulated quite big systems in the past ([~500 globular proteins](https://pubs.acs.org/doi/abs/10.1021/jp505809v)) using this level of detail. Another example is with [flexible proteins](https://doi.org/10.1017/qrd.2020.7) and for that publication there's also an example [input file](https://static.cambridge.org/content/id/urn:cambridge.org:id:article:S2633289220000071/resource/name/S2633289220000071sup001.pdf).

OpenMP pragmas should likely incorporated in `Energy::BasicEnergyAccumulator` and postpones energy summation until `operator double()` is called. That is, one would store a list of particle pairs (references) and then sum up their energies in a single OMP loop.

**Reply by mlund** [reply]
I just made a new branch, [`openmp-accumulator`](https://github.com/mlund/faunus/tree/openmp-accumulator) where you'll find a `OpenMPEnergyAccumulator`. It still needs the OMP pragmas and you should replace `BasicEnergyAccumulator` in `Nonbonded`. I suspect it will be advantegeous only for large systems.

**Reply by mlund** [reply]
Doing some testing, threading the energy summation using the C++17 parallel execution policy (`std::execution::par`) seems to outperform OpenMP. Implemented in above the above branch as well for further testing.

**Reply by directysj** [reply]
> In the implicit salt framework (Poisson-Boltzmann approximation; `yukawa`) it's normal to have a system net-charge as it's neutralized by implicit counterions, so there's no need to add counter-ions, unless you think their surface correlations are important. We have simulated quite big systems in the past ([~500 globular proteins](https://pubs.acs.org/doi/abs/10.1021/jp505809v)) using this level of detail. Another example is with [flexible proteins](https://doi.org/10.1017/qrd.2020.7) and for that publication there's also an example [input file](https://static.cambridge.org/content/id/urn:cambridge.org:id:article:S2633289220000071/resource/name/S2633289220000071sup001.pdf).
> 
> OpenMP pragmas should likely incorporated in `Energy::BasicEnergyAccumulator` and postpones energy summation until `operator double()` is called. That is, one would store a list of particle pairs (references) and then sum up their energies in a single OMP loop.

thank you for sharing these papers.

**Reply by directysj** [reply]
> Doing some testing, threading the energy summation using the C++17 parallel execution policy (`std::execution::par`) seems to outperform OpenMP. Implemented in above the above branch as well for further testing.

I am looking into it. Thanks a lot :)

**Reply by directysj** [reply]
When installing Faunus with MPI option to explore the 'parallel temper move' using command: cmake . -DENABLE_MPI=ON  
-DCMAKE_INSTALL_PREFIX:PATH="..."

This is giving error on make as: 

faunus.cpp: In function ‘void saveOutput(TimePoint&, docopt::Options&, Faunus::MetropolisMonteCarlo&, const Faunus::Analysis::CombinedAnalysis&)’:
faunus.cpp:372:30: error: reference to ‘MPI’ is ambiguous
     if (std::ofstream stream(MPI::prefix + args["--output"].asString()); stream) {
                              ^~~
In file included from faunus.cpp:3:
mpicontroller.h:31:15: note: candidates are: ‘namespace Faunus::MPI { }’
     namespace MPI {
               ^~~
Can I ask you for suggestions please?

**Reply by mlund** [reply]
> In file included from faunus.cpp:3:
> mpicontroller.h:31:15: note: candidates are: ‘namespace Faunus::MPI { }’
> namespace MPI {
> ^~~
> Can I ask you for suggestions please?

I cannot reproduce this with GCC10 nor with Clang using the master branch. What is your setup?

**Reply by directysj** [reply]
I am compiling master branch with GCC 8.3.0, CMAKE 3.20.0

**Reply by mlund** [reply]
hmm, I don't have gcc 8 on my machine, but try using `Faunus::MPI::prefix` in `faunus.cpp:372`.

**Reply by directysj** [reply]
this is not working.

**Reply by directysj** [reply]
Hello,

I have tried to install the Faunus on my laptop which has GCC 11.1.1, CLANG 12.0.0; OS is fedora 34. I am getting the same error messages with clang/GCC, as shown above. Unfortunately, GCC 10 local installation is ending up with errors too. Is it related to the mpicc, mpicxx, or mpirun installed on my computers? I am not getting any clue. Can I ask you help in this regard?

**Reply by directysj** [reply]
I re-downloaded using git cloning and replaced MPI::XYZ to Faunus::MPI::XYZ, and it works but an error message on $make install -----

CMake Error at examples/cmake_install.cmake:46 (file):
  file INSTALL cannot find
  "/home/jalim/softwares/mpiFaunus/faunus/examples/improper_dihedral/improper-dihedral.yml":
  No such file or directory.
Call Stack (most recent call first):
  cmake_install.cmake:57 (include)

make: *** [Makefile:136: install] Error 1

**Reply by directysj** [reply]
I just did $make clean and $ rm -fR CMakeCache.txt CMakeFiles _deps
Then, I copied improper_dihedral.yml to improper-dihedral.yml in that directory. It installed Faunus without errors. I will now check parallel temper and post here if any problem/update. 
Thank you!!

**Reply by mlund** [reply]
Thanks for finding that bug - fixed in 3b2acfa018573244061060f2084e32ca1399623e

**Reply by mlund** [reply]
In the meantime we have worked a little bit more in the parallel energy accumulator. Try `git pull; git checkout openmp-accumulator` and rebuild with `-DENABLE_TBB=on`. To enable the parallel scheme you'll GCC 10+ and add "parallel_policy: parallel" to the nonbonded section of the input. See `Energy::DelayedEnergyAccumulator` and `docs/schema.yml` for information about the input (documentation still not ready).

**Reply by mlund** [reply]
You can also enable `-DENABLE_OPENMP=on` and the `parallel_policy: openmp` should be available. In my tests `parallel` is faster than `openmp` and uses the C++17 parallel execution policy which why a new gcc in needed.

**Reply by mlund** [reply]
Sorry, it's `summation_policy` not ~~`parallel_policy`~~. Also, use `--verbosity 6` to get extra output. I suspect you'll need to go to large systems for this to give speedups. Any feedback on this is highly appreciated.

**Reply by directysj** [reply]
Thank you so much for all these suggestions and adding openmp .  I will check this out and let you know about parallel tempering as well as parallel/openmp .

**Reply by directysj** [reply]
I am testing parallel temper. 

I have tried temper.yml script, and titration.yml with adding temper move. 
The command is used as: yason.py temper.yml | mpirun --np 3 --stdin all faunus

It stuck after the command and nothing has happened, no errors. Only creating three files with names mpi0.stdout, mpi1.stdout, and  mpi2.stdout. I would like to get some suggestions ...

**Reply by directysj** [reply]
I have tried to compile `parallel energy accumulator` with `openmp` and `tbb` on. I am getting error followed by termination of `$make` .

faunus/_deps/tbb-src/src/test/test_mutex.cpp:44:10: fatal error: test/OpenMP_Mutex.h: No such file or directory
   44 | #include "test/OpenMP_Mutex.h"

Also several warnings with message: Deprecated Features appendix in the TBB reference manual

**Reply by mlund** [reply]
The parallel scheme doesn't require openmp, so perhaps try to compile with only `-DENABLE_TBB=on`. Also try to set `TBB_DIR` (see install section in manual) to instead specify a system installed TBB. I have not seen the error you describe, though, but we usually compile only on macos. Do you compile with `make faunus`? Btw. there's an updated version of the `openmp-accumulator` branch. It is more complete and the manual now explains how to use it (`docs/_docs/energy.md`).

**Reply by directysj** [reply]
I have tried without `openmp` option and included `TBB_DIR` path. It compiles and now I will test it.

**Reply by directysj** [reply]
Error while running with `summation_policy`: parallel in the `nonbonded` :

$ yason.py input.yml | faunus
energy -> nonbonded -> summation_policy -> 'parallel' is not of type 'array'

[faunus 19725] E: [json.exception.parse_error.101] parse error at line 1, column 1: syntax error while parsing value - unexpected end of input; expected '[', '{', or a literal
[faunus 19725] E: invalid input -> [json.exception.parse_error.101] parse error at line 1, column 1: syntax error while parsing value - unexpected end of input; expected '[', '{', or a literal
invalid input -> [json.exception.parse_error.101] parse error at line 1, column 1: syntax error while parsing value - unexpected end of input; expected '[', '{', or a literal

The section of input file where `summation policy` is added shown below:

```
energy:
    - bonded: {}
    - nonbonded:
         summation_policy: parallel 
         default:
              - lennardjones: {mixing: LB}
              - coulomb: {epsr: 78.7, type: yukawa, shift: true, debyelength: {{ debyelength }}, cutoff: 12}
         SER SER:
              - wca: {mixing: LB}
         SER THR:
              - wca: {mixing: LB}
```

**Reply by mlund** [reply]
> $ yason.py input.yml | faunus
> energy -> nonbonded -> summation_policy -> 'parallel' is not of type 'array'

It sounds as if `yason.py` is not using the new `docs/schema.yml` file. Can you try with `which yason.py` to see where the python script is located? Try running it from the source directory, i.e. something like `$HOME/faunus/scripts/yason.py`. The script should then look for `$HOME/faunus/scripts/../docs/schema.yml`

**Reply by directysj** [reply]
Thank you. Actually, I had installed it in a local directory `$HOME/localDir/` , where `faunus`, `yason.py` `ob2aam.py` were located. Then, I added the path to the .bashrc file. It was alright, but may be I had forgot to `$source ~/.bashrc`. It is working now (since I rebooted my laptop). I am sorry for it.

I will post it here about its performance by varying the system size.

**Reply by directysj** [reply]
Hello @mlund ,

I have tested simulations with ~20000 particles (beads) for 10,000 MC steps (`micro*macro`) . The simulations times are `2926, 1252, and 1036` minutes for` serial`, `TBB`, and `OpenMP` . It was running on `8 threads`. The total energy calculated from  all these runs is comparable.

**Reply by mlund** [reply]
Yes, it's not impressive scaling, but I'm working to improve this. Currently we're looking to see if SIMD vectorization using OpenMP can provide a speedup. I'll let you know when there's an update.

**Reply by directysj** [reply]
Thank you so much for sharing this info. As you are already working on cell lists for nonbonding forces, which will also speedup the code.

**Reply by directysj** [reply]
Hello @mlund ,

While installing `Faunus` with `GCC 10.3.0`, I am getting errors on `$make` as followings:

```
faunus/_deps/src/project_spdlog/include/spdlog/fmt/bundled/format.h: In member function 'void fmt::v6::internal::basic_writer<Range>::write(T, fmt::v6::internal::basic_writer<Range>::format_specs)':
faunus/_deps/src/project_spdlog/include/spdlog/fmt/bundled/format.h:1725:14: error: '__builtin_signbit' is not a member of 'std'; did you mean '__builtin_signbit'?
 1725 |     if (std::signbit(value)) {  // value < 0 is false for NaN so use signbit.
      |              ^~~~~~~
<built-in>: note: '__builtin_signbit' declared here
faunus/_deps/src/project_spdlog/include/spdlog/fmt/bundled/format.h:1732:15: error: '__builtin_isfinite' is not a member of 'std'; did you mean '__builtin_isfinite'?
 1732 |     if (!std::isfinite(value)) {
      |               ^~~~~~~~
<built-in>: note: '__builtin_isfinite' declared here
faunus/_deps/src/project_spdlog/include/spdlog/fmt/bundled/format.h:1733:23: error: '__builtin_isinf_sign' is not a member of 'std'; did you mean '__builtin_isinf_sign'?
 1733 |       auto str = std::isinf(value) ? (fspecs.upper ? "INF" : "inf")
      |                       ^~~~~
<built-in>: note: '__builtin_isinf_sign' declared here
```

However, with `GCC 11.+` it is installing without errors.

**Reply by directysj** [reply]
Hello @mlund 

I have installed new version (2.5.0) of `Faunus` with `GCC11+`, which is giving me errors when running the code:

`E: energy -> nonbonded -> error creating energy -> cannot allocate memory for energy pairs: std::bad_alloc`

However, it is giving errors while installing as shown above on cluster with `GCC10.3.0`.

Thanks
Jalim

**Reply by mlund** [reply]
Regarding the compilation error, I suspect the gcc10 version on the cluster uses an older version of the libstdc++, likely the system default which is typically old. The error comes from an auxiliary library and seems connected to missing c++11/14/17 features.
The other issue is due to memory allocation of the matrix for storing interaction pairs. If you have 20.000 pairs as before, I think it's trying to allocation 6.4 GB(!). I'm investigating if this is really needed. In any case, this should happen only when using non-serial summation policy, right?

**Reply by mlund** [reply]
676bb4ca3a0115f26c2596414ee7223fa4b1b02c reduces the memory requirement by a factor of two.

**Reply by mlund** [reply]
Please see PR #385 which will set an upper limit (~800 MB) for the pair list. Does this work on your system(?). You can run with `--verbosity 6` to get more information about what is being allocated.

**Reply by directysj** [reply]
> Regarding the compilation error, I suspect the gcc10 version on the cluster uses an older version of the libstdc++, likely the system default which is typically old. The error comes from an auxiliary library and seems connected to missing c++11/14/17 features.
> The other issue is due to memory allocation of the matrix for storing interaction pairs. If you have 20.000 pairs as before, I think it's trying to allocation 6.4 GB(!). I'm investigating if this is really needed. In any case, this should happen only when using non-serial summation policy, right?

Yes, you right about the memory allocation ERROR, I have reduced the number of particles in the system then it is gone. I was using non-serial summation. Regarding the compilation error, I will check it out.

**Reply by directysj** [reply]
For the same input script, I am getting numerous warnings `"W:humongous negative energy change"` in `2.5.0` version that is absent in `2.4.2` version, when `summation policy` is `parallel/openmp`; I have not tested it in `serial`. This is when I am using `GCC11+`.

**Reply by mlund** [reply]
Hopefully the large energy warning happens only during equilibration? That is, when loading from relaxed state file, it should be quiet. Perhaps it's more useful if the warning level is `debug` instead of `warn`(?)

**Reply by directysj** [reply]
> Hopefully the large energy warning happens only during equilibration? That is, when loading from relaxed state file, it should be quiet. Perhaps it's more useful if the warning level is `debug` instead of `warn`(?)

okay, I will test it with pre-equilibrated state file.

**Reply by directysj** [reply]
> Regarding the compilation error, I suspect the gcc10 version on the cluster uses an older version of the libstdc++, likely the system default which is typically old. The error comes from an auxiliary library and seems connected to missing c++11/14/17 features.

I could install `2.5.0` on cluster, finally. But, it is now showing warning that "`parallel/openmp summation policy is unavailable, falling back to serial`" . Looks like some libraries are still missing. It's running OK on my laptop which has `GCC11+` and TBB installed using root. On cluster, I installed `TBB` locally, `export LD_LIBRRARY_PATH` and path to the `TBB` config files.

**Reply by directysj** [reply]
Interestingly, If only `OPENMP` is `ON` (with `TBB OFF`), then also I am getting same warnings as 

```
[faunus 19057] W: 'openmp' summation unavailable; falling back to 'serial'
[faunus 19057] D: setting summation policy to "serial"
```
...........

**Reply by mlund** [reply]
Yes, I have seen this before.  I think it's because some versions of GCC require TBB for it's openmp support. The warning is triggered in `src/energy.h:1335` when `_OPENMP` (set by the compiler) is not defined.

**Reply by directysj** [reply]
> Yes, I have seen this before. I think it's because some versions of GCC require TBB for it's openmp support. The warning is triggered in `src/energy.h:1335` when `_OPENMP` (set by the compiler) is not defined.

Hello @mlund ,

I have installed `LLVM 12` (which gives `CLANG/CLANG++`) and Intel's `TBB`, locally. I have modified the `CMAKE` command from my previous build as:

`$CC=$HOME/llvm/bin/clang CXX=$HOME/llvm/bin/clang++ cmake -B build -DCMAKE_CXX_FLAGS=-stdlib=libc++ -DCMAKE_EXE_LINKER_FLAGS=-stdlib=libc++ -DENABLE_OPENMP=ON -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX:PATH="..." -DZLIB_LIBRARY=/usr/lib/x86_64-linux-gnu/libz.so` 

I am using `CMAKE 3.20` and `Anaconda3` for `python`.  Before CMAKE, I am exporting compiler flags as: 
`$export LDFLAGS="-L$HOME/llvm/lib $LDFLAGS"; export CFLAGS="-I$HOME/llvm/include $CFLAGS"; export CC=$HOME/llvm/bin/clang; export CXX=$HOME/llvm/bin/clang++`

Now, I can configure and make like shown below: 
........
```
[ 98%] Building CXX object src/CMakeFiles/faunus.dir/faunus.cpp.o
[100%] Linking CXX executable ../faunus
/usr/bin/ld: CMakeFiles/faunus.dir/faunus.cpp.o: in function `createProgressTracker(bool, unsigned int)':
faunus.cpp:(.text+0xcc6f): undefined reference to `ProgressIndicator::ProgressBar::ProgressBar(unsigned int, std::__1::basic_ostream<char, std::__1::char_traits<char> >&, unsigned int, char, char)'
```
...............................
And, then exit with `clang-12: error: linker command failed with exit code 1 (use -v to see invocation)`.

I don't have any idea, how to go from here. Can I ask you to suggest something? 
Thank you in advance!!

~Jalim Singh

**Reply by directysj** [reply]
> > Yes, I have seen this before. I think it's because some versions of GCC require TBB for it's openmp support. The warning is triggered in `src/energy.h:1335` when `_OPENMP` (set by the compiler) is not defined.
> 
> Hello @mlund ,
> 
> I have installed `LLVM 12` (which gives `CLANG/CLANG++`) and Intel's `TBB`, locally. I have modified the `CMAKE` command from my previous build as:
> 
> `$CC=$HOME/llvm/bin/clang CXX=$HOME/llvm/bin/clang++ cmake -B build -DCMAKE_CXX_FLAGS=-stdlib=libc++ -DCMAKE_EXE_LINKER_FLAGS=-stdlib=libc++ -DENABLE_OPENMP=ON -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX:PATH="..." -DZLIB_LIBRARY=/usr/lib/x86_64-linux-gnu/libz.so`
> 
> I am using `CMAKE 3.20` and `Anaconda3` for `python`. Before CMAKE, I am exporting compiler flags as:
> `$export LDFLAGS="-L$HOME/llvm/lib $LDFLAGS"; export CFLAGS="-I$HOME/llvm/include $CFLAGS"; export CC=$HOME/llvm/bin/clang; export CXX=$HOME/llvm/bin/clang++`
> 
> Now, I can configure and make like shown below:
> ........
> 
> ```
> [ 98%] Building CXX object src/CMakeFiles/faunus.dir/faunus.cpp.o
> [100%] Linking CXX executable ../faunus
> /usr/bin/ld: CMakeFiles/faunus.dir/faunus.cpp.o: in function `createProgressTracker(bool, unsigned int)':
> faunus.cpp:(.text+0xcc6f): undefined reference to `ProgressIndicator::ProgressBar::ProgressBar(unsigned int, std::__1::basic_ostream<char, std::__1::char_traits<char> >&, unsigned int, char, char)'
> ```
> 
> ...............................
> And, then exit with `clang-12: error: linker command failed with exit code 1 (use -v to see invocation)`.
> 
> I don't have any idea, how to go from here. Can I ask you to suggest something?
> Thank you in advance!!
> 
> ~Jalim Singh

Dear @mlund ,

I have looked the problem deeper and it seems that `cmake` is not passing some of library flags to the linker. This I found when `Cmake` `BOOL` option is kept on and in the detail it shows many times `-std=gnu++17` instead of `-std=c++17`.  Is this the problem of `cmake` OR the files related to `cmake` in this project?  Can I ask you to have a look at it? 

Thanking you in anticipation!!

**Reply by mlund** [reply]
The development environment can unfortunately be tricky to setup. I have not been able to reproduce the errors above. If nothing else works, I recommend to use the `scripts/Dockerfile` (see Installation about Docker in the manual). Alternatively, use Remote Containers with Visual Code Studio which will setup a working environment in a Docker image on all platforms that can run Docker (one can now click in a badge on the Faunus github site).
